File: 154. Find Minimum in Rotated Sorted Array II.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    if( nums == null || nums.length == 0 ) {
        return -1;
    }

    var start = 0;
    var end = nums.length - 1;
    var mid;

    while ( start + 1 < end ) {
        mid = start + Math.round( (end - start) / 2 );
        if ( nums[mid] < nums[end] ) {
            end = mid;
        } else if ( nums[mid] > nums[end] ) {
            start = mid;
        } else {
            end--;
        }
    }

    if ( nums[start] <= nums[end] ){
        return nums[start];
    } else {
        return nums[end];
    }
};

File: 154. Find Minimum in Rotated Sorted Array II.py
class Solution(object):
    def findMin(self, nums):
        if not nums or len(nums) == 0:
            return -1

        low = 0
        high = len(nums) - 1

        while low + 1 < high:
            mid = (low + high) / 2
            if nums[mid] < nums[high]:
                high = mid
            elif nums[mid] > nums[high]:
                low = mid
            else:
                high -= 1

        return min(nums[low], nums[high])
